xdsl/dialects/emitc.py

```diff
@@ -150,10 +181,42 @@ def is_integer_index_or_opaque_type(
     return _is_supported_integer_type(type_attr) or isa(type_attr, IndexType)
 
 
+def is_supported_emitc_type(type_attr: Attribute) -> bool:
+    """
+    Check if a type is supported by EmitC.
+    See [MLIR implementation](https://github.com/llvm/llvm-project/blob/main/mlir/lib/Dialect/EmitC/IR/EmitC.cpp#L62).
+    """
+    match type_attr:
+        case IntegerType():
+            return _is_supported_integer_type(type_attr)
+        case IndexType():
+            return True
+        case EmitC_ArrayType():
+            elem_type = cast(Attribute, type_attr.get_element_type())
+            return not isa(elem_type, EmitC_ArrayType) and is_supported_emitc_type(
+                elem_type
+            )
+        case Float16Type() | BFloat16Type() | Float32Type() | Float64Type():
+            return True
+        case TensorType():
+            elem_type = cast(Attribute, type_attr.get_element_type())
+            if isinstance(elem_type, EmitC_ArrayType):
+                return False
+            return is_supported_emitc_type(elem_type)
+        case TupleType():
+            return all(
```

Comment:
let's add a test for this

Reply:
added in 2c68a348651558d8d3ea6da3bbf02c6f4dedeef9